### PR TITLE
8214761: Bug in parallel Kahan summation implementation

### PR DESCRIPTION
**Pull request by igraves**
---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8214761](https://bugs.openjdk.java.net/browse/JDK-8214761): Bug in parallel Kahan summation implementation

### Reviewers
 * [Joe Darcy](https://openjdk.java.net/census#darcy) (@jddarcy - **Reviewer**) ⚠️ Review applies to 10b8dcda0646929a04fad9b508b232034234944d

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/4674/head:pull/4674` \
`$ git checkout pull/4674`

Update a local copy of the PR: \
`$ git checkout pull/4674` \
`$ git pull https://git.openjdk.java.net/jdk pull/4674/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 4674`

View PR using the GUI difftool: \
`$ git pr show -t 4674`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/4674.diff">https://git.openjdk.java.net/jdk/pull/4674.diff</a>

</details>

### CODE REVIEW

**Comment by theRealAph** [comment]
Crikey, how did we get that wrong?
It'd be nice if we had a regression test for this. Can you provide one, please?

File: src/java.base/share/classes/java/util/DoubleSummaryStatistics.java [hunk 1]
```diff
@@ -156,7 +156,9 @@ public void combine(DoubleSummaryStatistics other) {
         count += other.count;
         simpleSum += other.simpleSum;
         sumWithCompensation(other.sum);
-        sumWithCompensation(other.sumCompensation);
+
+        //Negating this value because low-order bits are in negated form
+        sumWithCompensation(-other.sumCompensation);
```
Comment (stefan-zobel) [inline src/java.base/share/classes/java/util/DoubleSummaryStatistics.java:161]: Wouldn't that be `double tmp =  sum + sumCompensation;` in getSum() in line 246 too?

**Comment by stefan-zobel** [comment]
What about `Collectors.computeFinalSum()` - should this be `double tmp = summands[0] + summands[1];` or `double tmp = summands[0] - summands[1];` ?

**Comment by stefan-zobel** [comment]
> Crikey, how did we get that wrong?
> It'd be nice if we had a regression test for this. Can you provide one, please?

I found this: https://mail.openjdk.java.net/pipermail/core-libs-dev/2018-December/057239.html
Ivan Gerasimov already tackled this back then. His webrev still exists and it contains a simple regression test.

File: src/java.base/share/classes/java/util/DoubleSummaryStatistics.java (hunk 1 above)
Comment (igraves) [reply src/java.base/share/classes/java/util/DoubleSummaryStatistics.java:161]: Good catch will review and make the change.

**Comment by igraves** [comment]
Circling back on this. I've worked in the test from Ivan Gerasimov's email back when. It includes some additional comparisons to prior approaches to assert improvements in error.

**Review by jddarcy — APPROVED** [review]
The code changes look fine, but IMO the comments should be re-worded a bit.

Rather text like

// Negating this value because low-order bits are in negated form

I suggest something like

// Subtract compensation bits

The main compensation loop also subtracts the compensation bits. A comment like "subtract compensation bits" makes the corrected handling of them seem less anomalous.

**Pull request closed without merging**